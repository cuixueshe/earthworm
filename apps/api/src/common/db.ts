import { Logger } from "@nestjs/common";
import { DefaultLogger, LogWriter } from "drizzle-orm";
import { drizzle } from "drizzle-orm/postgres-js";
import * as postgres from "postgres";

import { schemas } from "@earthworm/schema";

let connection: postgres.Sql;

async function createConnection() {
  return postgres(process.env.DATABASE_URL ?? "");
}

export async function endDB() {
  if (connection) {
    await connection.end();
    connection = null;
  }
}

export async function setupDB() {
  if (connection) return;

  const logger = new Logger("DB");

  class CustomDbLogWriter implements LogWriter {
    write(message: string) {
      logger.verbose(message);
    }
  }

  logger.debug(`Connecting to ${process.env.DATABASE_URL}`);
  logger.debug(`SECRET: ${process.env.SECRET}`);

  connection = await createConnection();

  return drizzle(connection, {
    schema: schemas,
    logger: new DefaultLogger({ writer: new CustomDbLogWriter() }),
  });
}
