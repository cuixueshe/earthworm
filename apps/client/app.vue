<template>
  <NuxtLayout>
    <HttpErrorProvider>
      <NuxtPage />
    </HttpErrorProvider>
  </NuxtLayout>
</template>

<script setup lang="ts">
import { useLogto } from "@logto/vue";
import { onMounted } from "vue";

import { useDarkMode } from "~/composables/darkMode";
import { isAuthenticated } from "~/services/auth";
import { useUserStore } from "./store/user";

const { initDarkMode } = useDarkMode();

async function setup() {
  const userStore = useUserStore();
  const logto = useLogto();

  if (isAuthenticated()) {
    const res = await logto.fetchUserInfo();
    userStore.initUser(res!);
  }
}

setup();

onMounted(() => {
  initDarkMode();
});
</script>

<style>
#jfToolbar,
.mod-json {
  display: none !important;
}
</style>
