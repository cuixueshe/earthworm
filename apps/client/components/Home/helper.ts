import { ref } from "vue";

import type { UserRecentCoursePackResponse } from "~/api/userCourseProgress";
import { fetchUserRecentCoursePacks } from "~/api/userCourseProgress";

const coursePacks = ref<UserRecentCoursePackResponse[]>([]);

export function useRecentCoursePack() {
  async function fetchCoursePacks() {
    coursePacks.value = await fetchUserRecentCoursePacks();
  }

  return {
    fetchCoursePacks,
    coursePacks,
  };
}
