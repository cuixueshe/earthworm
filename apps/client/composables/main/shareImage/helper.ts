const loadImage = async (url: string): Promise<HTMLImageElement> => {
  const imgEl = document.createElement("img");
  imgEl.src = url;
  return new Promise((resolve, reject) => {
    imgEl.onload = () => resolve(imgEl);
    imgEl.onerror = reject;
  });
};

export const initCanvas = () => {
  const canvasEl = document.createElement("canvas");
  canvasEl.width = 900;
  canvasEl.height = 1200;
  return canvasEl
}

export const convertSVGtoImg = async (
  svg: string,
  canvasEl: HTMLCanvasElement,
  fullFormat: string
) => {
  const dataHeader = "data:image/svg+xml;charset=utf-8";
  const encodeAsUTF8 = (str: string) =>
    `${dataHeader},${encodeURIComponent(str)}`;

  const svgData = encodeAsUTF8(svg);
  const img = await loadImage(svgData);
  
  canvasEl.getContext("2d")?.drawImage(img, 0, 0, 900, 1200);
  const dataURL = await canvasEl.toDataURL(fullFormat, 1.0);
  return dataURL;
};

export const copyImage = (canvasEl: HTMLCanvasElement, fullFormat: string) => {
  canvasEl?.toBlob((blob) => {
    if (blob) {
      navigator.clipboard.write([new ClipboardItem({ [blob.type]: blob })]).then(() => {
        console.log("Image copied.");
        return true
      });
    }
  }, fullFormat);
};

export const clearCanvas = (canvasEl: HTMLCanvasElement) => {
  const ctx = canvasEl.getContext('2d');
  ctx?.clearRect(0, 0, canvasEl.width, canvasEl.height);
}

export const fontEn = () => fetch(new URL('~/assets/fonts/EBGaramond-BoldItalic.ttf', import.meta.url)).then(
  (res) => res.arrayBuffer(),
);

export const fontZh = () => fetch(new URL('~/assets/fonts/nzgrKangxi.ttf', import.meta.url)).then(
  (res) => res.arrayBuffer(),
)
