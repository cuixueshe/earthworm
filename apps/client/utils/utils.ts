export function numberToChinese(n: number) {
  const units = ["", "十", "百", "千"];
  const chars = ["零", "一", "二", "三", "四", "五", "六", "七", "八", "九"];
  let num = n;
  if (num === 0) return chars[0];
  let str = "";
  let zeroFlag = false;

  for (let i = 0; num > 0; i++) {
    const digit = num % 10;
    num = Math.floor(num / 10);

    if (digit === 0) {
      if (!zeroFlag) {
        str = chars[digit] + str;
        zeroFlag = true;
      }
    } else {
      str = chars[digit] + units[i] + str;
      zeroFlag = false;
    }
  }
  // 10-19进行处理
  if (n >= 10 && n < 20) {
    str = str.slice(1);
  }
  // 去除末尾的零
  if (str.endsWith(chars[0])) {
    str = str.slice(0, -1);
  }
  return str;
}
