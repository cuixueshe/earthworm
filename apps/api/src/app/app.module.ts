import { Module } from '@nestjs/common';
import { GlobalModule } from '../global/global.mudule';
import { UserModule } from '../user/user.module';
import { AuthModule } from '../auth/auth.module';
import { CourseModule } from '../course/course.module';
import { UserProgressModule } from '../user-progress/user-progress.module';
import { ToolModule } from '../tool/tool.module';
import { RedisModule } from '@nestjs-modules/ioredis';
import { RankModule } from '../rank/rank.module';
import { GameModule } from '../game/game.module';
import { ScheduleModule } from '@nestjs/schedule';
import { CronJobModule } from '../cron-job/cron-job.module';

@Module({
  imports: [
    GlobalModule,
    UserModule,
    AuthModule,
    CourseModule,
    UserProgressModule,
    ToolModule,
    RankModule,
    GameModule,
    CronJobModule,
    RedisModule.forRootAsync({
      useFactory: () => ({
        type: 'single',
        url: process.env.REDIS_URL,
        password: process.env.REDIS_PASSWORD,
      }),
    }),
    ScheduleModule.forRoot(),
  ],
})
export class AppModule {}
