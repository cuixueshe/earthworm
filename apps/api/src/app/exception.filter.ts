import { ArgumentsHost, Catch, ExceptionFilter, HttpException } from "@nestjs/common";

@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse();
    const status = exception.getStatus();

    const exceptionResponse = exception.getResponse();
    const message = exception.message ?? `${status >= 500 ? "Service Error" : "Client Error"}`;

    const errorResponse = {
      data: {},
      message,
    };
    if (typeof exceptionResponse === "object" && exceptionResponse.hasOwnProperty("message")) {
      errorResponse.message = exceptionResponse["message"];
    }
    response.status(status);
    response.header("Content-Type", "application/json; charset=utf-8");
    response.send(errorResponse);
  }
}
