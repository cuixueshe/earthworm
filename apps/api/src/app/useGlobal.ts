import { INestApplication, ValidationPipe } from "@nestjs/common";

import { HttpExceptionFilter } from "./exception.filter";

export const appGlobalMiddleware = (app: INestApplication) => {
  app.useGlobalPipes(new ValidationPipe());
  app.useGlobalFilters(new HttpExceptionFilter());
};
