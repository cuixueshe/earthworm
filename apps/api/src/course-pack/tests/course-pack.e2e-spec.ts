import { INestApplication } from "@nestjs/common";
import { Test, TestingModule } from "@nestjs/testing";
import * as request from "supertest";

import { insertCourse, insertCoursePack, insertStatement } from "../../../test/fixture/db";
import { cleanDB, signin } from "../../../test/helper/utils";
import { AppModule } from "../../app/app.module";
import { appGlobalMiddleware } from "../../app/useGlobal";
import { endDB } from "../../common/db";
import { DB, DbType } from "../../global/providers/db.provider";

describe("course-pack e2e", () => {
  let app: INestApplication;
  let db: DbType;
  let token: string;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    appGlobalMiddleware(app);
    db = moduleFixture.get<DbType>(DB);

    await app.init();

    await cleanDB(db);

    token = await signin(moduleFixture);
  });

  afterEach(async () => {
    await cleanDB(db);
    await endDB();
    await app.close();
  });

  it("get: /course-pack", async () => {
    await insertCoursePack(db);
    return request(app.getHttpServer())
      .get("/course-pack")
      .set("Authorization", `Bearer ${token}`)
      .expect(200)
      .expect(({ body }) => {
        expect(body.length).toBe(1);
      });
  });

  it("get: /course-pack/:coursePackId", async () => {
    const { id: coursePackId } = await insertCoursePack(db);
    await insertCourse(db, coursePackId);
    await insertCourse(db, coursePackId);

    return request(app.getHttpServer())
      .get(`/course-pack/${coursePackId}`)
      .set("Authorization", `Bearer ${token}`)
      .expect(200)
      .expect(({ body }) => {
        expect(body).toHaveProperty("courses");
        expect(body).toHaveProperty("order");
        expect(body).toHaveProperty("title");
      });
  });

  it("get: /course-pack/:coursePackId/courses/:courseId", async () => {
    const { id: coursePackId } = await insertCoursePack(db);
    const { id: courseId } = await insertCourse(db, coursePackId);
    await insertStatement(db, courseId, 1);
    await insertStatement(db, courseId, 2);

    return request(app.getHttpServer())
      .get(`/course-pack/${coursePackId}/courses/${courseId}`)
      .set("Authorization", `Bearer ${token}`)
      .expect(200)
      .expect(({ body }) => {
        expect(body).toHaveProperty("coursePackId");
        expect(body).toHaveProperty("order");
        expect(body).toHaveProperty("statements");
      });
  });

  it("get: /course-pack/:coursePackId/courses/:courseId/next", async () => {
    const { id: coursePackId } = await insertCoursePack(db);

    const { id: courseId } = await insertCourse(db, coursePackId, {
      order: 1,
      title: "第一课",
    });
    const { id: courseIdNext } = await insertCourse(db, coursePackId, {
      order: 2,
      title: "第二课",
    });

    return request(app.getHttpServer())
      .get(`/course-pack/${coursePackId}/courses/${courseId}/next`)
      .set("Authorization", `Bearer ${token}`)
      .expect(200)
      .expect(({ body }) => {
        expect(body.id).toBe(courseIdNext);
      });
  });

  it("post: /course-pack/:coursePackId/courses/:courseId/complete", async () => {
    const { id: coursePackId } = await insertCoursePack(db);

    const { id: courseId } = await insertCourse(db, coursePackId, {
      order: 1,
      title: "第一课",
    });
    const { id: courseIdNext } = await insertCourse(db, coursePackId, {
      order: 2,
      title: "第二课",
    });

    return request(app.getHttpServer())
      .post(`/course-pack/${coursePackId}/courses/${courseId}/complete`)
      .set("Authorization", `Bearer ${token}`)
      .expect(201)
      .expect(({ body }) => {
        expect(body.nextCourse.id).toBe(courseIdNext);
      });
  });
});
