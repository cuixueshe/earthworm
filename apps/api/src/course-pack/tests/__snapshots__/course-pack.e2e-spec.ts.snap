// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`course-pack e2e get: /course-pack 1`] = `[]`;

exports[`course-pack e2e get: /course-pack/:coursePackId 1`] = `
{
  "courses": [
    {
      "completionCount": 0,
      "coursePackId": 1,
      "id": 1,
      "title": "第一课",
    },
    {
      "completionCount": 0,
      "coursePackId": 1,
      "id": 2,
      "title": "第一课",
    },
  ],
  "description": "这是一个课程包",
  "id": 1,
  "isFree": true,
  "title": "课程包",
}
`;

exports[`course-pack e2e get: /course-pack/:coursePackId/courses/:courseId 1`] = `
{
  "coursePackId": 1,
  "id": 1,
  "statementIndex": 0,
  "statements": [
    {
      "chinese": "你好",
      "courseId": 1,
      "english": "hello",
      "order": 1,
      "soundmark": "nihao",
    },
    {
      "chinese": "你好",
      "courseId": 1,
      "english": "hello",
      "order": 2,
      "soundmark": "nihao",
    },
  ],
  "title": "第一课",
}
`;
